src/ssh/allowed_signer.rs

```diff
@@ -0,0 +1,218 @@
+
+    /// Time at or before which the key is valid.
+    /// This is a standard option.
+    pub valid_before: Option<u64>,
+
+    /// Public key of the entry.
+    pub key: PublicKey,
+}
+
+/// A type which represents a collection of allowed signer entries.
+/// Please refer to [ssh-keygen-1.ALLOWED_SIGNERS] for more details about the format.
+/// [ssh-keygen-1.ALLOWED_SIGNERS]: https://man.openbsd.org/ssh-keygen.1#ALLOWED_SIGNERS
+#[derive(Debug, PartialEq, Eq)]
+pub struct AllowedSigners {
+    /// A collection of allowed signers
+    pub allowed_signers: Vec<AllowedSigner>,
+}
+
+impl AllowedSigner {
+    /// Parse an allowed signer entry from a given string.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use sshcerts::ssh::AllowedSigner;
+    ///
+    /// let allowed_signer = AllowedSigner::from_string(concat!(
+    ///     "user@domain.tld ecdsa-sha2-nistp384 AAAAE2VjZHNhLXNoYTItbmlzdHAzODQAAAAIbmlzdHAzODQAAABhBGJe",
+    ///     "+IDGRhlQdDp+/AIsTXGaVhWaQUbHJwqLDlQIh7V4xatO6E/4Uva+f70WzxgM7xHPGUqafqNAcxxVBP4jkx3HVDRSr7C3",
+    ///     "NVBpr0ZaKXu/hFiCo/4kry4H5MGMEvKATA=="
+    /// )).unwrap();
+    /// println!("{:?}", allowed_signer);
+    /// ```
+    pub fn from_string(s: &str) -> Result<AllowedSigner> {
+        let mut parts: VecDeque<&str> = s.split_whitespace().collect();
+
+        // An allowed signer must contian at least 3 parts: principals, key type and pubkey data
+        if parts.len() < 3 {
+            return Err(Error::InvalidFormat);
+        }
+
+        let principals = parts.pop_front().ok_or(Error::InvalidFormat)?;
+        let principals: Vec<&str> = principals.split(',').collect();
+        let principals = principals.iter().map(|s| s.to_string()).collect();
+
+        let key_data = parts.pop_back().ok_or(Error::InvalidFormat)?;
+        let kt = parts.pop_back().ok_or(Error::InvalidFormat)?;
+        let key = PublicKey::from_string(format!("{} {}", kt, key_data).as_str())?;
+
+        let mut cert_authority = false;
+        let mut namespaces = None;
+        let mut valid_after = None;
+        let mut valid_before = None;
+
+        for option in parts {
+            let option = option.to_lowercase();
+            let (key, value) = match option.split_once('=') {
+                Some(v) => v,
+                None => (option.as_str(), ""),
+            };
+            match key {
+                "cert-authority" => cert_authority = true,
+                "namespaces" => {
+                    if namespaces.is_some() {
+                        return Err(Error::InvalidFormat);
+                    }
+                    let namespaces_inner: Vec<&str> = value.trim_matches('"')
+                            .split(',')
+                            .collect();
+                    namespaces = Some(
+                        namespaces_inner.iter()
+                            .map(|s| s.to_string())
+                            .collect()
+                    );
+                },
```

Comment:
TODO: Rework the option parsing logic:
- Support spaces in double quotes. Not sure how this works for a namespace. Need to find the specs for namespace.
- Might need to enforce a format for namespace.

Reply:
> The key used for signing is specified using the -f option and may refer to either a private key, or a public key with the private half available via [ssh-agent(1)](https://man.openbsd.org/ssh-agent.1). An additional signature namespace, used to prevent signature confusion across different domains of use (e.g. file signing vs email signing) must be provided via the -n flag. Namespaces are arbitrary strings, and may include: “file” for file signing, “email” for email signing. For custom uses, it is recommended to use names following a NAMESPACE@YOUR.DOMAIN pattern to generate unambiguous namespaces.

From https://man.openbsd.org/ssh-keygen.1